aten/src/ATen/native/Normalization.cpp

```diff
@@ -627,32 +668,45 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cpu(
   const Tensor& running_mean = *running_mean_maybe_owned;
   const Tensor& running_var = c10::value_or_else(running_var_opt, [] {return Tensor();});
 
-  return AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "batch_norm_update_stats_cpu", [&] {
-      return batch_norm_cpu_update_stats_template<scalar_t, Var>(self, running_mean, running_var, momentum, 0);
+  int64_t n_input = self.size(1);
+  int64_t n = self.numel() / n_input;
+  const int64_t ndim = self.dim();
+
+  // Reduce all dimensions except dim=1
+  DimVector reduce_dims(ndim - 1);
+  reduce_dims[0] = 0;
+  for (const auto i : c10::irange(2, ndim)) {
+    reduce_dims[i - 1] = i;
+  }
+
+  Tensor save_mean = at::mean(self, /*dims=*/reduce_dims);
+  Tensor save_var_transform = at::empty({n_input}, self.options());
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "batch_norm_update_stats_cpu", [&] {
+      batch_norm_cpu_update_stats_template<scalar_t, Var>(self, running_mean, running_var, momentum, 0, save_mean, save_var_transform);
     });
+  return std::make_tuple(save_mean, save_var_transform);
 }
 
-std::tuple<Tensor, Tensor, Tensor> batch_norm_cpu(const Tensor& self, const c10::optional<Tensor>& weight_opt, const c10::optional<Tensor>& bias_opt, const c10::optional<Tensor>& running_mean_opt, const c10::optional<Tensor>& running_var_opt,
-                                                  bool train, double momentum, double eps) {
-  // See [Note: hacky wrapper removal for optional tensor]
-  c10::MaybeOwned<Tensor> weight_maybe_owned = at::borrow_from_optional_tensor(weight_opt);
-  const Tensor& weight = *weight_maybe_owned;
-  const Tensor& bias = c10::value_or_else(bias_opt, [] {return Tensor();});
-  const Tensor& running_mean = c10::value_or_else(running_mean_opt, [] {return Tensor();});
-  const Tensor& running_var = c10::value_or_else(running_var_opt, [] {return Tensor();});
-
-  checkBackend("batch_norm_cpu", {self, weight, bias, running_mean, running_var}, Backend::CPU);
```

Comment:
@bdhirsh I am suspicious that omitting this check is the reason for the last CI failure i'm still facing.  

https://www.internalfb.com/intern/opensource/ci/job/log/3080152672308172/
2021-11-09T00:36:51.5070547Z AssertionError: "Could not run 'aten::native_batch_norm' with arguments from the 'QuantizedCPU'" does not match "Could not run 'aten::empty.memory_format' with arguments from the 'QuantizedCPU' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build). If you are a Facebook employee using PyTorch on mobile, please visit https://fburl.com/ptmfixes for possible resolutions. 'aten::empty.memory_format' is only available for these backends: [CPU, CUDA, Meta, MkldnnCPU, SparseCPU, SparseCUDA, SparseCsrCPU, SparseCsrCUDA, BackendSelect, Python, Named, Conjugate, Negative, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, AutogradXLA, AutogradLazy, AutogradXPU, AutogradMLC, AutogradHPU, AutogradNestedTensor, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, Tracer, UNKNOWN_TENSOR_TYPE_ID, Autocast, Batched, VmapMode, Functionalize].


It looks like the test (test_fusion.py:90) is expecting a particular failure of native_batch_norm, but that's no longer happening and empty is failing instead.

        with self.assertRaisesRegex(RuntimeError, "Could not run 'aten::native_batch_norm' with arguments from the 'QuantizedCPU'"):
            checkQuantized(model)

I think you had suggested I shouldn't use checkBackend in a device-specific way in the meta-kernel before. But is it OK to have it check specifically for unallowed devices such as quantized?  Or what would be the best way to handle this?

Reply:
ok- do you recommend modifying the unit test then?  or should I try to raise the same error as previously raised from inside meta::native_batch_norm?

Reply:
That means that the error message for quantized ops will get a little worse though - we could probably make it better, but that would require some special handling in the codegen.

Reply:
Yeah, let me take a look and see how hard it would be to add the check in the codegen. I think it would be reasonable to add it if isn't too much of a hack.

Reply:
https://github.com/pytorch/pytorch/pull/71928 should improve the error message (I tried rebasing your changes on top of it and making the error message no longer mentions `empty.memory_format`).

You'll still need to tweak the test a bit because I couldn't easily use the exact same error message though.

Reply:
https://www.internalfb.com/diff/D33823417 just landed internally, so you should be able to rebase and (fingers crossed) the error will go away

Reply:
yea i have already rebased the internal diff and verified it works.  (Unfortunately, I'm still stuck on a separate internal error)